feat(render): add Hats module

minecrrrr opened the pull request:
### **Description**

This module adds decorative hats to the client. Currently, it includes a customizable ChinaHat mode implemented using the WorldRenderEnvironment API.

### **Screenshots**

**Third-person**
<img width="1920" height="1080" alt="2025-12-26_12 45 32" src="https://github.com/user-attachments/assets/f50678c0-a916-499f-b94d-ecd5a7b805f8" />

<img width="1920" height="1080" alt="2025-12-26_12 46 37" src="https://github.com/user-attachments/assets/872b3224-d696-4b6c-b0c5-d214ddcbe4da" />

**First-person**

First Person View (on)

<img width="1920" height="1080" alt="2025-12-26_12 46 56" src="https://github.com/user-attachments/assets/9ebbda7e-0aeb-4b2b-a027-5d587c19a2af" />

First Person View (off)

<img width="1920" height="1080" alt="2025-12-26_12 47 00" src="https://github.com/user-attachments/assets/643e6c4b-af22-46c8-b650-e93669112cac" />

### **Settings**

<img width="241" height="482" alt="изображение" src="https://github.com/user-attachments/assets/ca73f0ed-87e4-4c12-b3ae-9002d14e71da" />

minecrrrr committed:
feat (render): add Hats module

This module adds decorative hats to the client. Currently, it includes a customizable ChinaHat mode implemented using the WorldRenderEnvironment API.

minecrrrr committed:
refactor: rename setting to First Person View

minecrrrr commented:
### Note:
 The NoWeb module shown in the screenshot is part of my local test environment and is not included in this Pull Request. This PR is strictly about the new Hats module

minecrrrr committed:
fix(render): improve HatsMode generic typing

MukjepScarlet commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/render/hats/modes/HatsChinaHat.kt:
```diff
@@ -0,0 +1,106 @@
+    private val height by float("HeightOffset", 0.1f, 0f..1f)
+
+    private val colors = tree(Color())
+    private class Color : Configurable("Color") {
```
`private val color by color("Color", Color4b(0, 0, 255, 125))`

minecrrrr committed:
refactor(render): use native Color delegate in ChinaHat

MukjepScarlet committed:
refactor

MukjepScarlet commented:
I want to rename it because it can be both dunce cap and bamboo hat lol

MukjepScarlet commented:
And I have seen another type of 'hat'... Probably will be added later.

MukjepScarlet committed:
indent

minecrrrr commented:
That sounds great! I'm glad this mode opened up some ideas for more hats. Looking forward to seeing what you'll add!

minecrrrr commented:
Regarding your suggestions — no problem at all! I'm totally fine with the changes and I actually appreciate your refactor. Let's go with 'ConeHat' then, as it’s more universal.
By the way, I'm already working on a 'Halo' (angel ring) mode, so I'll probably submit another PR soon!

minecrrrr commented:
Should I rename it myself or will you do it?

MukjepScarlet commented:
I will do

minecrrrr commented:
Ok

MukjepScarlet committed:
rename to 'ConeHat'

minecrrrr committed:
Merge branch 'CCBlueX:nextgen' into nextgen

MukjepScarlet committed:
simplify

MukjepScarlet commented:
I think current code can do a simple halo too, just need some more parameters. Please check. @minecrrrr

minecrrrr commented:
I'm actually almost done with the Halo mode.

minecrrrr commented:
I also created a separate RenderUtils for the module. It will be added along with the Halo mode.

minecrrrr committed:
Fixe(ConeHat): parameter names in HatsConeHat

minecrrrr commented:
I fixed the issues in HatsConeHat. The error was in the parameter names: they should be outerColor4b and innerColor4b , not outerColor and innerColor. @MukjepScarlet

minecrrrr committed:
Revert parameter names in ConeHat to fix CI build failure

minecrrrr commented:
I apologize, it seems my local environment/IDE works differently, and I mistakenly tried to apply a fix to your code

MukjepScarlet commented:
Can I add some **big** refactors or merge it?

minecrrrr commented:
Yep. By the way, I've already finished the halo. Should I add it to this PR?

minecrrrr commented:
You can refactor it if you want

MukjepScarlet commented:
You can add that because I believe the refactor can make it conflict

minecrrrr committed:
feat(Hats): Add Halo and Orbs modes, enhance render utilities

This PR introduces three new highly configurable modes for the new **Hats** module: Cone, Halo and Orbs. All modes feature custom 3D rendering with wave and rotation animations.
Key Changes:

    **Added HatsHalo.kt** (Torus model).
    **Added** HatsOrbs.kt (Animated 3D-Rhombus "Sims style" model).
    **Introduced HatsRenderUtils.kt** for shared rendering logic.
    Implemented **full configuration via ClickGUI** (Speed, Size, Wave, Rotation).

minecrrrr committed:
refactor(Hats): rename ConeHat mode to Cone

minecrrrr committed:
Merge branch 'CCBlueX:nextgen' into nextgen

minecrrrr committed:
docs(Hats): add comments to Hats modes

minecrrrr committed:
Merge branch 'nextgen' of https://github.com/minecrrrr/LiquidBounce into nextgen

minecrrrr committed:
docs(Hats): add comments to Hats modes

minecrrrr committed:
docs(Hats): Modify comments for Hats modes

minecrrrr committed:
docs(Hats): Modify comments for Hats modes

minecrrrr committed:
feat(Hats): add new Hat - "Flower"

minecrrrr committed:
fix(Hats): add "Halo", "Orbs" and "Flower" modes to ModuleHat.kt

minecrrrr commented:
wtf

minecrrrr committed:
fix(Hats): fix unresolved reference error in getStarRadius

minecrrrr committed:
refactor(Hats): split long rendering lines in HatsOrbs.kt

MukjepScarlet commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/render/hats/HatsRenderUtils.kt:
```diff
@@ -0,0 +1,49 @@
+import kotlin.math.sin
+import kotlin.math.abs
+
+fun getAngle(i: Int, segments: Int) = i * Math.PI * 2 / segments
```
use UnitCircle

minecrrrr replied:
UnitCircle doesn't really fit for 3D or toroidal shapes, which is what I'm working on right now, as it only handles 2D circular logic, while toroids require two radii.

MukjepScarlet replied:
Use radius = 1, then the x and y becomes `cos(theta)` and `sin(theta)`. See how `drawGradientCircle` does.

minecrrrr commented:
._.

minecrrrr commented:
I can't fix everything right now. I need to head out for about 3 hours.

MukjepScarlet committed:
format code

minecrrrr commented:
Hello, @MukjepScarlet . I wanted to clarify a few things:
    **New content**: Is it okay if I commit new modes for Hats (Star), update HatsRenderUtils.kt, and add an update for the Flower mode? I’ve got these ready to go!
    **ClickGUI Bug**: I also noticed a bug in the ClickGUI where switching between modes doesn't refresh the settings/options immediately. They only update after reopening the GUI. Do you have any pointers on where I should look to fix this, or should I open a separate issue for it?

Flower:

https://github.com/user-attachments/assets/f06375df-3820-4697-9a0f-e7be58de56cc

<img width="1920" height="1009" alt="2025-12-27_21 38 15" src="https://github.com/user-attachments/assets/6b82054c-1c47-4985-8ee2-d76752f7d4ad" />


Star:

https://github.com/user-attachments/assets/1553bbba-352f-44a8-8b28-837881546649

<img width="1920" height="1009" alt="2025-12-27_21 38 37" src="https://github.com/user-attachments/assets/efbe189e-c65f-42ed-af6e-d5868f476157" />

Bug: 

https://github.com/user-attachments/assets/5e5b942c-315f-4d72-9b13-a0d0e0301897

MukjepScarlet commented:
You can add what you want.
The ClickGUI bug is a known issue https://github.com/CCBlueX/LiquidBounce/issues/7275

MukjepScarlet commented:
Mark as `ready` when you think you finish it.

minecrrrr commented:
I'm about to push an update, and then the code will be ready for merge.

minecrrrr committed:
feat(Hats): Update Hats module and add new modes

**Added** HatsStar.kt (new Star mode)
**Updated** HatsRenderUtils.kt (improved rendering logic and meshes)
**Updated** HatsFlower.kt (mode improvements)

minecrrrr committed:
Merge branch 'nextgen' of https://github.com/minecrrrr/LiquidBounce into nextgen

minecrrrr committed:
Update Hats module and add new modes

**Added** HatsStar.kt
**Updated** HatsRenderUtils.kt
**Updated** HatsFlower.kt
**Updated** HatsOrbs.kt (add @author)
**Updated** HatsHalo.kt (add @author)

minecrrrr commented:
The module is **ready for merge**. I'm always down to open another PR with even more hats later if you're interested. ;)

minecrrrr committed:
Cleanup(Hats): Removed broken rotation logic in HatsStar.kt

Removed experimental head-tracking code that was causing rendering issues in third-person view

minecrrrr commented:
The code is **definitely** ready for merge now

minecrrrr commented:
Sorry, I've realized that a few things need to be touched up. I'll be done in a moment

MukjepScarlet committed:
Merge branch 'nextgen' into nextgen

minecrrrr committed:
refactor(Hats): add gradients, improved code style

**Added** gradient support for better visuals. (HatsConeHat.kt, HatsFlower.kt, HatsHalo.kt)
**Refactored** code structure to comply with the 120-character line length limit. (HatsStar.kt)
**Added** new gradient calculation logic. (HatsRenderUtils.kt)

minecrrrr committed:
Merge branch 'nextgen' of https://github.com/minecrrrr/LiquidBounce into nextgen

minecrrrr committed:
Refactor: Increased segment count for smoother gradients

minecrrrr committed:
Refactor(Hats): Increased segment count for smoother gradients

minecrrrr commented:
...

MukjepScarlet commented:
Read what it said. reformat

minecrrrr commented:
Ok

minecrrrr commented:
Since I'm already fixing the build issues, I'd like to include some extra polish and new features as well. Is that okay with you?

minecrrrr committed:
refactor(Hats): fix & refactor Hats Module

minecrrrr committed:
Merge branch 'nextgen' of https://github.com/minecrrrr/LiquidBounce into nextgen

minecrrrr commented:
I've fixed the build issues and linting errors. @MukjepScarlet

minecrrrr committed:
Feat: add angular gradient support to ConeHat

minecrrrr committed:
refactor(HatsStar.kt): Removed useless setting

minecrrrr committed:
Refactor: unified segment count across all hat modes for consistent gradients

minecrrrr commented:
Sorry for constantly switching the PR back to Draft. This time, I’ve definitely polished everything up

MukjepScarlet commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/render/hats/modes/HatsHalo.kt:
```diff
@@ -0,0 +1,141 @@
+ */
+internal object HatsHalo : HatsMode("Halo") {
+
+    private val height by float("HeightOffset", 0.2f, 0f..1f)
```
Move duplicated stuff to the parent class

minecrrrr replied:
I've moved showInFirstPerson, height, and Colors to the parent class to fully eliminate code duplication as requested. While this means all hats now share a global color scheme, it keeps the codebase much cleaner. If we need individual colors back later, we can refactor it then

minecrrrr replied:
I’ve unresolved the conversation because the proposed refactor (moving colors to the parent class) is technically incompatible with the current engine. Moving the property delegates (by color) away from the mode objects breaks the GUI binding — sliders in the menu stop updating the variables in real-time. I’ve reverted the colors to the per-mode structure to restore functionality and ensure the module actually works. This is a technical limitation of how the config system handles property registration.

MukjepScarlet commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/render/hats/modes/HatsOrbs.kt:
```diff
@@ -0,0 +1,149 @@
+        val pos = player.interpolateCurrentPosition(it.partialTicks)
+
+        renderEnvironmentForWorld(it.matrixStack) {
+            withPositionRelativeToCamera(pos.add(0.0, player.bbHeight + height.toDouble(), 0.0)) {
```
Same here. create an abstract function to render and create `handler` on parent class

minecrrrr replied:
I tried moving the renderHat logic to the parent class, but it's not feasible with the current architecture. The drawCustomMesh function is a DSL-based extension that requires a specific rendering context from renderEnvironmentForWorld. When moved to an abstract method in a separate class, it loses this context and breaks. To avoid over-complicating the rendering pipeline with hacky workarounds, I've kept the rendering logic within the module's scope.

MukjepScarlet commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/render/hats/HatsRenderUtils.kt:
```diff
@@ -0,0 +1,129 @@
+}
+
+data class TorusQuad(
+    val p1: Triple<Float, Float, Float>,
```
Vector3f/Vec3f

minecrrrr commented:
I tried moving the renderHat logic to the parent class, but it's not feasible with the current architecture. The drawCustomMesh function is a DSL-based extension that requires a specific rendering context from renderEnvironmentForWorld. When moved to an abstract method in a separate class, it loses this context and breaks. To avoid over-complicating the rendering pipeline with hacky workarounds, I've kept the rendering logic within the module's scope.

minecrrrr committed:
Refactor(Hats): Refactor HatsRenderUtils and move shared settings to parent class.

**HatsRenderUtils**: Changed data types from Triple to Vector3f for better performance and cleaner code.
**Settings Refactor**: Moved height and showInFirstPerson to the parent class to remove code duplication in hat modes.

minecrrrr committed:
Refactor(Hats): Consolidate color settings into the main module class.

**Colors Refactor**: Moved the Colors configuration object from individual hat modes to the main ModuleHats class.
**Code Cleanup**: Successfully eliminated remaining duplicated code across the module, ensuring all hat modes now use a unified color configuration.
**Centralized Control**: This change simplifies future updates to the color system and follows the request to move shared logic to the parent/main scope.

minecrrrr committed:
Revert "Consolidate color settings into the main module class"

Reason: Reverting the centralized color configuration due to a break in the property delegate binding.

Technical issue: Moving color settings to the parent class/main module caused the GUI sliders to lose connection with the variables. Changes in the menu are no longer reflected during rendering.

Resolution: Restored per-mode color objects to ensure stable configuration and functional UI.

minecrrrr commented:
@MukjepScarlet, I've tried to move the rendering into an abstract function in the parent class, but drawCustomMesh simply doesn't work outside of the main module. It keeps throwing errors because it's tied to the specific environment of the main class. If I move it, the code just turns red and won't compile. If you know a way to make this work, please show me an example, because right now it seems impossible to separate them

MukjepScarlet committed:
refactor: move common parameters

MukjepScarlet committed:
refactor: move common draw actions

MukjepScarlet committed:
refactor: VertexConsumer helper function

minecrrrr committed:
Fix: Wrap single-line statements in curly braces

minecrrrr committed:
Fix(Hats): color spinSpeed functionality

Resolved an issue where the spinSpeed parameter was not correctly applied to color modules. The rotation/cycling speed now works as intended.

minecrrrr committed:
fix & refactor(Hats): refined Hats settings

HatsStar.kt: Lowered sharpness limit for finer adjustment.

HatsCone.kt: Fixed a bug where ColorSpin setting wa not showing up in ClickGUI.

minecrrrr committed:
refactor(Hats): Simplify variable names and access

minecrrrr committed:
refactor(Hats): split HatsRenderUtils into separate files within utils package

minecrrrr committed:
Update package-lock.json

minecrrrr committed:
Deep architectural refactor: unified rendering logic and complexity optimization

Complete overhaul of the Hats module structure (refactored 90% of files).

Standardized geometry calculations: Moved core logic to TorusUtils.kt and parent HatsMode.

Eliminated Code Bloat: Replaced redundant logic with unified calculation methods.

Linter compliance: Fixed all Complexity and LongParameterList issues across the entire module.

Data Integrity: Introduced comprehensive data classes for cleaner parameter passing.

Optimized performance by reducing redundant object allocations in render loops.

minecrrrr committed:
style: fix linter issues and cleanup

minecrrrr commented:
I've completed a deep architectural overhaul. Most of the logic has been extracted to the parent class to ensure clean and maintainable code. I’ve also fixed all complexity issues, and local detekt confirms a 'BUILD SUCCESSFUL', so it's ready for review

<img width="1919" height="1030" alt="изображение" src="https://github.com/user-attachments/assets/30841d2d-79a3-4631-bedd-2c76dcbe7b2d" />

minecrrrr commented:
@1zun4, I was really hoping to see this module in the 0.36.0 release, which is why I've been working so hard to polish everything before the December 31st deadline.
As @MukjepScarlet  knows, we’ve already gone through several iterations of refactoring to make the code perfectly clean. Now the linter is happy, there are no conflicts, and my local builds are working flawlessly. The community is already hyped for these new hats (Orbs, Halo, etc.).
Given that the module is fully stable and finished, is there any chance we could include it in 0.36.0 as a New Year's gift for the users? It would be a shame to make them wait another month for a feature that is already 100% ready.

1zun4 commented:
I see, but don't worry. You will still theoretically end up with v0.36.0 because the v0.36.0 milestone has been renamed v0.35.3. I think 1.21.11 is more stable than 1.21.8 by now, so it makes sense to release it as an early patch version and include all the new features in the actual 0.36.0.

Just probably not this year anymore. Sorry about that.

minecrrrr commented:
Ah, all good
Yeah, I was just hoping to make it for the Dec 31st deadline as a New Year's gift, but no big deal if it's for the 'actual' 0.36.0 release. Thanks for the clarification and the update!

localpthebest commented:
hats 👍

MukjepScarlet committed:
reformat code

MukjepScarlet committed:
move function + reformat

MukjepScarlet committed:
same code cleanup, use float for render

minecrrrr commented:
@MukjepScarlet, I’ve addressed all the requested changes and refactored the code. Could you please take another look at the PR and dismiss the Change Request if everything looks good now

MukjepScarlet commented:
Further refactors needed. If I have time I will make them ready in this weekend.

minecrrrr committed:
Feat(Hats): Add "Friend View" option

minecrrrr commented:
https://github.com/user-attachments/assets/f0240b42-d71d-4f16-927c-61b5bb7f59e9

minecrrrr committed:
feat(Hats): add "Render Distance" option and fix detekt warnings

minecrrrr committed:
feat(hats): add "Show Damage" option and separate friend options

minecrrrr committed:
refactor(Hats): fix detekt warnings

minecrrrr committed:
refactor(Hats): fix detekt warnings

1zun4 commented:
> 2025-12-31.130200.online-video-cutter.com.mp4

That is a great idea.

minecrrrr commented:
Thanks:)

xDeadForMeLife commented:
hello do you have a gas stove at home?

minecrrrr commented:
Do you have soap and rope?

MukjepScarlet committed:
refactor: merge code

MukjepScarlet committed:
Color settings cleanup

MukjepScarlet commented:
Still too many copy-paste code but I need to sleep

minecrrrr committed:
fix(Hats): hat visibility in FreeLook while in first-person

minecrrrr commented:
I'm not sure whose fault it is, but Orbs is totally broken now

minecrrrr committed:
fix(Hats): repair Orbs rendering logic

minecrrrr commented:
The rotation logic was broken as well...

minecrrrr commented:
Mistake

minecrrrr commented:
@MukjepScarlet, visuals are complete and ready for review whenever you're available. I don't mean to be pushy, just wanted to keep you updated

minecrrrr committed:
Merge branch 'CCBlueX:nextgen' into nextgen

SenkJu commented:
This ia definitely something that would belong in the fun category.

MukjepScarlet commented:
Both fun and render are acceptable to me

1zun4 commented:
Belongs into Render.

minecrrrr commented:
Is there anything I need to do with RenderShortcuts.kt right now?

MukjepScarlet commented:
I'll solve the conflict (probably)

MukjepScarlet commented:
Because I rewrote some render code yesterday.

MukjepScarlet committed:
Merge branch 'nextgen' into nextgen

MukjepScarlet committed:
import cleanup

MukjepScarlet committed:
move util to abstract mode

MukjepScarlet committed:
spinSpeed toggle -> zero value

MukjepScarlet reviewed (approved):
Anyway we can improve it in the future (if needed)

minecrrrr commented:
Thanks for the review and merge.
Sounds good, we can improve it later if needed

minecrrrr commented:
@MukjepScarlet Currently, Hats conflicts with Target Render in KillAura (Ghost), causing visual artefacts. Can I revert merge?


https://github.com/user-attachments/assets/9f192783-a75d-41cb-ad10-cd3e6ce45951

minecrrrr commented:
I would be happy to address all the issues right now, but I need to get some rest to avoid being late for school

MukjepScarlet commented:
This is probably another issue. Textured render + normal render.

MukjepScarlet commented:
@minecrrrr This issue has been fixed in #7656
